### PR TITLE
[JIT] Add basic aliasing checks for tensor inputs

### PR DESCRIPTION
**Pull request by ttanpcs**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79661
* __->__ #79474

- Added aliasing checks in SchemaCheckMode that raise a RuntimeError if an argument aliases an output that is not specified by the matching schema
- Added python bindings for isAliasOf: a method that checks if two tensors are aliases of each other, before_set: which returns the before sets and after_set: which returns the after sets.
- Created IncorrectAliasTensor as a subclass of torch.Tensor that purposefully aliases incorrectly
- Tested that alias checks in SchemaCheckMode work for both correct and incorrect situations

**Note that the cases of inputs aliasing with each other, outputs aliasing with each other, and cases where the before and after sets are not equivalent (this should only occur when the after set is the wildcard set) are not implemented. These cases will be addressed in a later pr.**

### CODE REVIEW

File: test/jit/test_schema_check.py [hunk 1]
```diff
@@ -96,3 +142,26 @@ def test_mutation_check_fail_multiple_operators(self):
...
+            y = torch.rand((3, 3), requires_grad=True)
+            with enable_torch_dispatch_mode(SchemaCheckMode()):
+                IncorrectAliasTensor(x).add(IncorrectAliasTensor(y), alpha=2)
+
+    def test_alias_check_fail_multiple_operators(self):
+        with self.assertRaises(RuntimeError):
+            x = torch.rand((3, 3), requires_grad=True)
+            y = torch.rand((3, 3), requires_grad=True)
+            with enable_torch_dispatch_mode(SchemaCheckMode()):
+                IncorrectAliasTensor(x).sin().relu().add(IncorrectAliasTensor(y), alpha=2)
```
Comment (davidberard98) [inline test/jit/test_schema_check.py:159]: nit: just for completeness could you add a test where add is one of the ops in the middle?

File: torch/csrc/jit/python/init.cpp [hunk 2]
```diff
@@ -1498,9 +1498,34 @@ void initJITBindings(PyObject* module) {
             const AliasInfo* aliasInfo = self.alias_info();
             return aliasInfo && aliasInfo->isWrite();
           })
+      .def_property_readonly(
```
Comment (davidberard98) [inline torch/csrc/jit/python/init.cpp:1501]: nit: since we have so many properties that we want to get from AliasInfo, maybe we should bind the AliasInfo type and then add these properties onto that, just so that the python code doesn't look too far off from the c++ code (you can do this as a separate PR if you prefer)

File: torch/testing/_internal/schema_check_mode.py [hunk 3]
```diff
@@ -37,9 +44,17 @@ def has_mutated(before, after):
         for argument in func._schema.arguments:
             name = argument.name if argument.name != "self" else "input"
             if arguments.get(name) is not None:
-                before = tree_flatten(arguments.get(name))[0]
-                after = tree_flatten(cloned_arguments.get(name))[0]
+                before = tree_flatten(cloned_arguments.get(name))[0]
+                after = tree_flatten(arguments.get(name))[0]
                 if (any([has_mutated(i, j) for i, j in zip(before, after)]) and not argument.is_mutable):
                     raise RuntimeError(f"Argument {name} is not defined as mutable but was mutated")
+                for i in before:
```
Comment (davidberard98) [inline torch/testing/_internal/schema_check_mode.py:51]: very minor nit: i feel like `i` is usually the name of an index, maybe rename it `b`? totally not important though, you can ignore if you don't want to change it.

File: torch/testing/_internal/schema_check_mode.py [hunk 4]
```diff
@@ -37,9 +44,17 @@ def has_mutated(before, after):
...
+                    if not isinstance(out, tuple):
+                        if (have_values_aliased(i, out) and not is_aliasing(argument, func._schema.returns[0])):
+                            raise RuntimeError(f'Argument {name} is not defined to alias output but was aliasing')
```
Comment (davidberard98) [inline torch/testing/_internal/schema_check_mode.py:54]: I think you should remove this check, ops are sometimes marked as aliasing when they only alias *sometimes*.

For example, `aten::to`.

```python
x = torch.rand((2, 2), dtype=torch.float)
y = x.to(dtype=torch.float)
z = x.to(dtype=torch.half)
```

^ in this example, I would expect that x and y are aliasing, but x and z are not aliasing.

Alternative option is to log when this happens but not error out, that way you might be able to identify ops that say they are aliasing but actually *never* alias.

File: test/jit/test_schema_check.py [hunk 5]
```diff
@@ -16,6 +18,50 @@
                        "\tpython test/test_jit.py TESTNAME\n\n"
                        "instead.")
 
+# This TorchDispatchTensor Subclass is used to test incorrectly aliasing ops
+# This is done by aliasing the output to the first argument for all ops in INCORRECT_OPS
+
+class IncorrectAliasTensor(torch.Tensor):
```
Comment (davidberard98) [inline test/jit/test_schema_check.py:24]: awesome! This could be pretty useful in the future too if we get around to fixing batch_norm (in which case we would need to find a new way to test mutation)

**Review by davidberard98 — COMMENTED** [review]
Overall looks good! Just had one major comment, and a few minor suggestions

File: torch/csrc/jit/python/init.cpp (hunk 2 above)
Comment (goldenxuett) [reply torch/csrc/jit/python/init.cpp:1501]: Sure! I'll do that in a separate PR just so this one doesn't get too cluttered.

File: torch/testing/_internal/schema_check_mode.py (hunk 4 above)
Comment (goldenxuett) [reply torch/testing/_internal/schema_check_mode.py:54]: From what I understand, this check currently works for ops that are marked as aliasing when they only alias sometimes because it only raises an error if two values are aliasing **and** the arguments are expected to not be aliasing. 

For the case you mentioned above, the fact that x and z are not aliasing would not raise an error since the values are not aliasing and an error is only raised when values are aliasing. Though, I could definitely log out these other cases as you mentioned. If I wanted to do that what would I use to log?

Comment (davidberard98) [reply torch/testing/_internal/schema_check_mode.py:54]: Whoops! Sorry, misread this line (thought it was the opposite check) - it looks good.

File: test/jit/test_schema_check.py [hunk 6]
```diff
@@ -16,6 +18,50 @@
                        "\tpython test/test_jit.py TESTNAME\n\n"
                        "instead.")
 
+# This TorchDispatchTensor Subclass is used to test incorrectly aliasing ops
+# This is done by aliasing the output to the first argument for all ops in INCORRECT_OPS
```
Comment (bdhirsh) [inline test/jit/test_schema_check.py:22]: minor nit: the comment says that this subclass is used to test incorrectly aliasing ops, but technically it's `SchemaCheckMode` that's responsible for that. It looks like the main purpose of this class is to test that `SchemaCheckMode` behaves as expected, by "simulating" an op with an incorrect schema and asserting that the mode picks up the error.

Do you mind updating the description a bit?

Can you make it a bit clearer in this comment that the purpose of this class is basically as a test case for `schema_check_mode`? It doesn't look like it's act

File: test/jit/test_schema_check.py [hunk 7]
```diff
@@ -96,3 +142,35 @@ def test_mutation_check_fail_multiple_operators(self):
                 x.tanh_()
                 x.relu_()
                 batch(SchemaCheckMode(x))
+
+    # Tests that an exception is raised for a mismatching alias
+    def test_alias_check_fail(self):
+        with self.assertRaises(RuntimeError):
```
Comment (bdhirsh) [inline test/jit/test_schema_check.py:148]: nit: for best practice, can you use `self.assertRaisesRegex(RuntimeError, "my expected error message")`?

File: torch/testing/_internal/schema_check_mode.py [hunk 8]
```diff
@@ -23,6 +24,21 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
...
             return not torch.equal(before, after) if isinstance(before, torch.Tensor) and isinstance(after, torch.Tensor) else False
 
+        def have_values_aliased(lhs, rhs):
+            return torch._C._is_alias_of(lhs, rhs) if isinstance(lhs, torch.Tensor) and isinstance(rhs, torch.Tensor) else False
+
+        def is_aliasing(lhs_argument, rhs_argument):
+            return bool(len(lhs_argument.before_set & rhs_argument.before_set))
+
+        def unwrap(e):
+            if isinstance(e, torch.Tensor) and not type(e) == torch.Tensor:
```
Comment (davidberard98) [inline torch/testing/_internal/schema_check_mode.py:34]: nit: could you add a comment about why you need this? (e.g. to handle IncorrectAliasTensor)

**Review by davidberard98 — APPROVED** [review]
LGTM! just had one comment, please add that and then you can merge once CI is green
